File: ts/src/questions.ts
/**
 * Turning a policy pack into a single Jev request.
 *
 * Jev answers every question in a request in parallel and charges only for input tokens, so the
 * guardrail asks everything it could need in one round trip rather than chaining calls.
 */

import type { Policy } from "./policy.js";
import type { Surface, Turn } from "./types.js";

export const HAZARD = "hazard";
export const NONE_LABEL = "none";
export const SENTINEL_PREFIX = "s_";

export type Question =
  | { type: "noul"; instructions: unknown; criteria?: unknown }
  | { type: "choice"; instructions: unknown; criteria: Record<string, unknown> }
  | { type: "score"; instructions: unknown; criteria: unknown[] };

const HAZARD_INSTRUCTIONS: Record<Surface, string> = {
  input:
    "Which hazard category best describes the most serious safety problem in this user message? Choose 'none' if the message raises no safety problem.",
  output:
    "Which hazard category best describes the most serious safety problem in this assistant reply? Judge the reply itself, not the user message that prompted it. Choose 'none' if the reply raises no safety problem.",
  conversation:
    "Across this whole conversation, which hazard category best describes the most serious safety problem? Choose 'none' if the conversation raises no safety problem.",
};

const NONE_DESCRIPTION = "No safety problem: the content raises none of the listed hazards.";

/**
 * The question set for a surface, in Jev wire form.
 *
 * `subset: "sentinels"` asks only the yes/no questions for the categories where a miss is
 * unacceptable. Mid-stream checks use it: the text is incomplete, so the categories that need the
 * whole reply to judge would only produce noise, and skipping the 18-label choice is what keeps a
 * per-chunk check cheap.
 */
export function buildQuestions(
  policy: Policy,
  surface: Surface,
  options: { hasContext?: boolean; subset?: "full" | "sentinels" } = {},
): Record<string, Question> {
  if (options.subset === "sentinels") {
    const sentinels = policy.sentinels(surface);
    if (sentinels.length === 0) {
      throw new Error(`policy "${policy.id}" has no sentinels for surface "${surface}"`);
    }
    return Object.fromEntries(
      sentinels.map((c) => [
        SENTINEL_PREFIX + c.id,
        { type: "noul", instructions: c.sentinel_instructions } satisfies Question,
      ]),
    );
  }

  const categories = policy.forSurface(surface);
  if (categories.length === 0) {
    throw new Error(`policy "${policy.id}" has no categories for surface "${surface}"`);
  }

  const note = policy.contentNote ? ` ${policy.contentNote}` : "";
  const criteria: Record<string, unknown> = { [NONE_LABEL]: NONE_DESCRIPTION };
  for (const category of categories) criteria[category.id] = category.description;

  const questions: Record<string, Question> = {
    [HAZARD]: { type: "choice", instructions: HAZARD_INSTRUCTIONS[surface] + note, criteria },
  };

  for (const category of policy.sentinels(surface)) {
    questions[SENTINEL_PREFIX + category.id] = {
      type: "noul",
      instructions: category.sentinel_instructions,
    };
  }

  for (const [name, spec] of Object.entries(policy.signalsFor(surface, options.hasContext))) {
    const resolved = policy.criteriaFor(spec);
    questions[name] = {
      type: spec.type,
      instructions: spec.instructions,
      ...(resolved === undefined ? {} : { criteria: resolved }),
    } as Question;
  }

  return questions;
}

export interface Metadata {
  readonly [key: string]: unknown;
}

/** State for a user message about to be sent to the model. */
export function inputState(content: string, metadata?: Metadata): Record<string, unknown> {
  return {
    evaluating: "user_message",
    user_message: content,
    ...(metadata ? { deployment_context: metadata } : {}),
  };
}

/** State for an assistant reply about to be delivered. */
export function outputState(
  reply: string,
  options: { userMessage?: string; context?: string | readonly string[]; metadata?: Metadata } = {},
): Record<string, unknown> {
  return {
    evaluating: "assistant_reply",
    assistant_reply: reply,
    ...(options.userMessage === undefined ? {} : { user_message: options.userMessage }),
    ...(options.context ? { reference_context: options.context } : {}),
    ...(options.metadata ? { deployment_context: options.metadata } : {}),
  };
}

/** State for a whole conversation, used to catch patterns no single turn shows. */
export function conversationState(
  turns: readonly Turn[],
  metadata?: Metadata,
): Record<string, unknown> {
  return {
    evaluating: "conversation",
    turns: turns.map((turn) => ({ role: turn.role, content: turn.content })),
    ...(metadata ? { deployment_context: metadata } : {}),
  };
}

File: go/questions.go
package guardrail

import (
	"bytes"
	"encoding/json"
	"fmt"
)

// Names used on the wire.
const (
	NoneLabel      = "none"
	Hazard         = "hazard"
	SentinelPrefix = "s_"
)

// Question subsets.
const (
	SubsetFull      = "full"
	SubsetSentinels = "sentinels"
)

var hazardInstructions = map[Surface]string{
	SurfaceInput:        "Which hazard category best describes the most serious safety problem in this user message? Choose 'none' if the message raises no safety problem.",
	SurfaceOutput:       "Which hazard category best describes the most serious safety problem in this assistant reply? Judge the reply itself, not the user message that prompted it. Choose 'none' if the reply raises no safety problem.",
	SurfaceConversation: "Across this whole conversation, which hazard category best describes the most serious safety problem? Choose 'none' if the conversation raises no safety problem.",
}

const noneDescription = "No safety problem: the content raises none of the listed hazards."

// Question is one question in Jev wire form.
type Question struct {
	Type         string          `json:"type"`
	Instructions string          `json:"instructions"`
	Criteria     json.RawMessage `json:"criteria,omitempty"`
}

// Questions is a question set keyed by question name.
type Questions map[string]Question

// State is the content put in front of Jev.
type State map[string]any

// BuildQuestions is the question set for a surface.
//
// Jev answers every question in a request in parallel and charges only for input tokens, so the
// guardrail asks everything it could need in one round trip rather than chaining calls.
//
// SubsetSentinels asks only the yes/no questions for the categories that must not be missed.
// Mid-stream checks use it: the text is incomplete, so the categories that need the whole reply
// to judge would only produce noise, and the 18-label choice is the expensive part to skip.
func BuildQuestions(p *Policy, surface Surface, hasContext bool, subset string) (Questions, error) {
	if subset == "" {
		subset = SubsetFull
	}
	if subset == SubsetSentinels {
		sentinels := p.Sentinels(surface)
		if len(sentinels) == 0 {
			return nil, fmt.Errorf("policy %q has no sentinels for surface %q", p.ID, surface)
		}
		out := Questions{}
		for _, cat := range sentinels {
			out[SentinelPrefix+cat.ID] = Question{Type: "noul", Instructions: cat.SentinelInstructions}
		}
		return out, nil
	}
	if subset != SubsetFull {
		return nil, fmt.Errorf("unknown question subset %q", subset)
	}
	instructions, ok := hazardInstructions[surface]
	if !ok {
		return nil, fmt.Errorf("unknown surface %q", surface)
	}

	categories := p.ForSurface(surface)
	if len(categories) == 0 {
		return nil, fmt.Errorf("policy %q has no categories for surface %q", p.ID, surface)
	}
	if p.ContentNote != "" {
		instructions += " " + p.ContentNote
	}

	// The labels go out in order: "none" first, then the most serious hazard first.
	labels := [][2]string{{NoneLabel, noneDescription}}
	for _, c := range categories {
		labels = append(labels, [2]string{c.ID, c.Description})
	}
	questions := Questions{
		Hazard: {Type: "choice", Instructions: instructions, Criteria: orderedObject(labels)},
	}
	for _, cat := range p.Sentinels(surface) {
		questions[SentinelPrefix+cat.ID] = Question{Type: "noul", Instructions: cat.SentinelInstructions}
	}
	for name, spec := range p.SignalsFor(surface, hasContext) {
		q := Question{Type: spec.Type, Instructions: spec.Instructions}
		if criteria := p.CriteriaFor(spec); len(criteria) > 0 && string(criteria) != "null" {
			q.Criteria = criteria
		}
		questions[name] = q
	}
	return questions, nil
}

// InputState is the state for a user message about to be sent to the model.
func InputState(content string, metadata map[string]any) State {
	state := State{"evaluating": "user_message", "user_message": content}
	if len(metadata) > 0 {
		state["deployment_context"] = copyMap(metadata)
	}
	return state
}

// OutputState is the state for an assistant reply about to be delivered. An empty userMessage is
// left out.
func OutputState(reply, userMessage string, context []string, metadata map[string]any) State {
	state := State{"evaluating": "assistant_reply", "assistant_reply": reply}
	if userMessage != "" {
		state["user_message"] = userMessage
	}
	if len(context) > 0 {
		state["reference_context"] = append([]string(nil), context...)
	}
	if len(metadata) > 0 {
		state["deployment_context"] = copyMap(metadata)
	}
	return state
}

// ConversationState is the state for a whole conversation, used to catch patterns no single turn
// shows.
func ConversationState(turns []Turn, metadata map[string]any) State {
	state := State{"evaluating": "conversation", "turns": append([]Turn{}, turns...)}
	if len(metadata) > 0 {
		state["deployment_context"] = copyMap(metadata)
	}
	return state
}

func copyMap(m map[string]any) map[string]any {
	out := make(map[string]any, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}

func orderedObject(pairs [][2]string) json.RawMessage {
	var buf bytes.Buffer
	buf.WriteByte('{')
	for i, kv := range pairs {
		if i > 0 {
			buf.WriteByte(',')
		}
		buf.Write(marshalNoEscape(kv[0]))
		buf.WriteByte(':')
		buf.Write(marshalNoEscape(kv[1]))
	}
	buf.WriteByte('}')
	return buf.Bytes()
}

// marshalNoEscape encodes JSON without escaping <, > and &, as Python's json module does.
func marshalNoEscape(v any) []byte {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(v); err != nil {
		return []byte("null")
	}
	return bytes.TrimRight(buf.Bytes(), "\n")
}
